### ISSUE TITLE
Session Handler fails to unserialize when Pipes exist in session data

### ISSUE THREAD

**Issue by ooglek** [ADOdb/ADOdb#537]
A decision was made in 2004 to use some code in the `ADODB_Session` handler that would split a `urldecode()`'d string on a pipe `|`. This was before there was an option to use the raw serialize/unserialize via `session.serialize_handler = php_serialize` which was added in PHP 5.4 or so. The default is still to use `php` which has this problem.

If one uses a pipe character in a session, the session cannot be unserialized as the array results in items that are not valid serialized strings. 

The session module assumes but does not check to see what serialize_handler is configured and assumes `session.serialize_handler = php`.

A few options:

1. Use `session_decode()`. This is safest because it handles all the different handlers quietly. Requires saving current session data to a temp variable while the ADODB code does stuff with it and replacing the session back after it's done extracting. Would need to see if using session_decode() would wipe out `$_SESSION` during any code execution outside of the current op. Don't love this but it future-proofs the implementation.

2. Use [this handy recursive snippet](https://gist.github.com/phred/1201412) from @phred. I timed it on a quad-core Xeon CPU E3-1231 v3 @ 3.40GHz with 32GB RAM and a 3,515 byte url-decoded string with a mix of arrays and pipes and whatnot. Elapsed: 0.0000479221. Probably fast enough. Plus no REGEX and no $_SESSION var replacement. BUT would need to check to see what serialize_handler was configured and support `php, php_binary and php_serialize`.

3. Something else!

`Ben` saw this problem on February 18, 2005 stating:

```
Sorry guys,
I'm also looking for such a function since a while.
This one doesn't work correctly neither.
Have a look at this code :

session_start();
$_SESSION['test'] = 'foo|bar';
print_r($_SESSION);
print_r(adodb_unserialize(session_encode()));

The results are incorrect.
To test a function, try passing it strings containing the following characters: |{};" (not in root keys, in which it's impossible for such a function to handle characters different than [0-9a-zA-Z_], due to a wanted php sessions' weakness because we need sometimes to register the $_SESSION array keys as global variables)

I cannot find one function that knows how to handle all datas correctly, not cool for storing binary datas :(

Ben
```

Original Discussion: https://web.archive.org/web/20160310025540/http://phplens.com/lens/lensforum/msgs.php?id=9821

**Comment by ooglek** [comment]
Argh. This still exists. I'm frustrated! It happened to me again this evening.

**Comment by ooglek** [comment]
Hey @dregad @mnewnham Would you be willing to take a look at this again? A pipe `|` anywhere in the session data breaks adodb Session Handling.

Maybe it is time to switch to something like JSON without the `urlencode()`? Or use serialize/unserialize without the preg_split?

Mysql/Mariadb both support JSON string queries now, though MySQL has native JSON column support and MariaDB only supports it as a string.

I don't know why things are stored the way they are, maybe someone can enlighten me? 

https://github.com/ADOdb/ADOdb/blob/e3e3d14051dcd4d9203a12de87848a43a0687da4/session/adodb-session2.php#L40-L48

**Comment by dregad** [comment]
Sorry, this kind of fell off the radar.

You submitted PR #538 as a proposed fix for this back then, but I don't think I actually ever reviewed it. Since the branch has conflicts now, maybe you can rebase to latest master and resolve those ? 

Alternatively, based on your last comment, maybe you'd prefer to come up with another, JSON-based fix ? This is probably OK too, but should not rely on DB-specific features to maintain maximum cross-platform compatibility. Switching to JSON will probably also require some code to handle the transition from the legacy serialization.

**Comment by ooglek** [comment]
Agreed, I will make it as PHP-centric and as widely supported as possible. I think allowing the user to configure to use JSON rather than legacy would make sense, though if it works, after a few releases it might make sense to use JSON as the default to help future users avoid this decade-old deficiency! 😁

**Comment by mnewnham** [comment]
Given that session.serialize_handler now supports PHP standard serialize/unserialize, I think our focus should be on moving the product to that, rather than anything else custom. That would allow us to remove the adodb_unserialize() methods completely.

**Comment by ooglek** [comment]
I'm down with that! Anything I can do to help?

**Comment by ooglek** [comment]
Sounds like it fell off the radar again? Any progress on moving to JSON or using the standard PHP serialize functions?

**Comment by ooglek** [comment]
Hi, so this is still a problem, and no changes have been made. What are your thoughts @mnewnham @dregad -- I'm happy to contribute a change or fix, but I'd like to know it'll get considered, and the best path forward as well.

**Comment by dregad** [comment]
@ooglek thanks for your message. I'm sorry but I actually was under the impression that *you* were working on this, following the exchange in late 2022... Expecting you to either update your PR #538 as mentioned in https://github.com/ADOdb/ADOdb/issues/537#issuecomment-1243409544, or propose another approach.

Your contribution would be welcome and of course considered, this is a community-driven project.

**Comment by mnewnham** [comment]
My own approach has been to do a complete rewrite of the sessions package, to take advantage of various new features of PHP 

- the new serialization methods
- improved driver support
- Overcoming the limited database connection support in the original version

The code is also tied into support for monolog logging which I've been using for a while now.

If you are interested in looking at my approach you can find it in t[his repo](https://github.com/ADOdb/ADOdb/tree/master-future-cache-extensions)  You can find all the code for the sessions management in the _**/dist**_ directory.

**Comment by ooglek** [comment]
I CAN work on it! Just wanted to check in so as not to duplicate efforts or design something that flies in the face of ADODB ethos, or at least coding standards!

I will try to review @mnewnham's repo and see what I can make of it.

For the session portion, are there any goals, known issues, or design criteria? E.g. must support binary as well as structured PHP data? Must use native PHP code where possible, etc.

I know what MY needs are, but as ADODB is likely in broad use, any new session implementation for encoding would need to work in most-to-all current situations as well, and I am not well versed with how other consumers of the library are using the session2 package.

I will noodle with @mnewnham's inspiration and see if I can come up with something, propose my approach and hopefully get some feedback, then write code.
